### PR TITLE
Validation workflow on card

### LINKED ISSUE #1149: Improve validation workflow on card
When the group name is empty during group creation, Catima shows a message "Group name cannot be empty" right at the UI element where it's wrong:
image

It would be good if it did this for the card creation/edit view too.

### PR DESCRIPTION
**Pull request by TomerPacific**
Fixes #1149 

Added logic and string literal, called **_card_name_is_empty_**, to allow to show the hint in the edit text when creating/editing a card.

Translated the value to almost all languages.

catima

catima

catima2

### CODE REVIEW

**Comment by TheLastProject** [comment]
Okay, I finally have some energy to look at stuff again. Thanks for your patience :)

Looking at this, it is a good start but there are few small things:

1. Could you please use `getError()` to make it so the check isn't duplicated like in https://github.com/CatimaLoyalty/Android/blob/1779aef162f1b03ada856e21ec24d426e3957498/app/src/main/java/protect/card_locker/ManageGroupsActivity.java#L161-L166?
2. Please only add translations you actually speak, I want to avoid Google Translate level translations.
3. I guess noStoreError can go with this new translation
4. `cardId` and `validBalance` also get checked in `doSave`, it would be nice if both of those also used this `setError` method to keep the UI consistent

**Comment by TomerPacific** [comment]
@TheLastProject , hope you had a nice vacation.
Wanting to verify the points you listed:

1. Do you want me to use getError on storeFieldEdit inside the onTextChanged?
2. I'll revert all the languages I don't speak (kind of wish I didn't invest all the time to translate it for this)
3. I don't understand what the relevance of noStoreError has to do with group name verification?

**Comment by TheLastProject** [comment]
> 1. Do you want me to use getError on storeFieldEdit inside the onTextChanged?

No, inside the `onSave` as that is what checks if everything is valid.

> 2. I'll revert all the languages I don't speak (kind of wish I didn't invest all the time to translate it for this)

Thanks :sweat_smile: 

> 3. I don't understand what the relevance of noStoreError has to do with group name verification?

It doesn't, but it seems ugly to me if in the edit view of the 3 elements that can have an invalid value only 1 ever uses the red exclamation point design. It can also confuse users if different fields on the same screen reacts in a different way to having an invalid value. I guess we can skip that but I don't want to release an update if the behaviour is inconsistent so I'll have to do the other 2 fields myself then before release :)

**Comment by TomerPacific** [comment]
@TheLastProject , thanks for the clarifications.

I usually think it is best to keep the PR relevant to the issue and it isn't quite apparent that you want other things to be done in the same PR.

I'll work on those changes and post a comment here when I'm done.

**Comment by TomerPacific** [comment]
@TheLastProject , I have made the changes you requested.
Regarding the build failing, it is doing that on the unit test phase.
When I ran the specific unit tests that failed locally, they passed without an issue.

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java [hunk 1]
```diff
@@ -409,9 +414,8 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
                 try {
                     BigDecimal balance = Utils.parseBalance(s.toString(), tempLoyaltyCard.balanceType);
                     updateTempState(LoyaltyCardField.balance, balance);
-                    validBalance = true;
                 } catch (ParseException e) {
-                    validBalance = false;
+                    balanceField.setError(getString(R.string.parsingBalanceFailed, balanceField.getText().toString()));
```
Comment (TheLastProject) [inline app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:418]: Looking at your screenshot the message doesn't fit below the box so we definitely need to simplify this string. Maybe just remove it and say "Invalid balance"

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java [hunk 2]
```diff
@@ -880,7 +888,13 @@ public void onResume() {
...
         }
 
-        if (cardId != null) {
+        CharSequence cardFieldTextViewError = cardIdFieldView.getError();
+
+        if (cardFieldTextViewError != null) {
+            Snackbar.make(cardIdFieldView,
+                    cardFieldTextViewError.toString(),
+                    Snackbar.LENGTH_LONG).show();
+        } else if (cardId != null) {
```
Comment (TheLastProject) [inline app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:897]: I think it's this part of code that's causing it, but there's definitely something broken now.

1. Start Catima
2. Click +
3. Click Manually enter ID
4. Enter "abc" in Card ID
5. Click "Aztec"
6. Card ID field is empty and 2 errors pop up

image

File: app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -145,6 +145,7 @@
     </plurals>
     <string name="group_name_already_in_use">Group name already in use</string>
     <string name="group_name_is_empty">Group name cannot be empty</string>
+    <string name="card_name_is_empty">Card name cannot be empty</string>
```
Comment (TheLastProject) [inline app/src/main/res/values/strings.xml:148]: The old message "No name entered" doesn't seem to appear anywhere anymore, so should be removed.

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java [hunk 4]
```diff
@@ -1315,7 +1331,9 @@ private void doSave() {
             return;
         }
 
-        if (!validBalance) {
+        CharSequence balanceFieldEditTextError = balanceField.getError();
+
+        if (balanceFieldEditTextError != null) {
             Snackbar.make(balanceField, getString(R.string.parsingBalanceFailed, balanceField.getText().toString()), Snackbar.LENGTH_LONG).show();
```
Comment (TheLastProject) [inline app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:1337]: This is not using the `balanceFieldEditTextError.toString()` like the rest.

**Review by TheLastProject — CHANGES_REQUESTED** [review]

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java (hunk 4 above)
Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:1337]: @TheLastProject , the string value **parsingBalanceFailed** has a placeholder for the value, so I figured you would want to use it. 

` <string name="parsingBalanceFailed"><xliff:g>%s</xliff:g> does not seem to be a valid balance.</string>`

If not, then we need to use a different string from the strings.xml.

Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:1337]: I don't understand your comment.

On line 418 you use `setError()` to put the full string in it already:
```
balanceField.setError(getString(R.string.parsingBalanceFailed, balanceField.getText().toString()));
```

So shouldn't you be possible to just do this?
```
Snackbar.make(balanceField, balanceFieldEditTextError.toString(), Snackbar.LENGTH_LONG).show();
```

Or am I missing something?

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java (hunk 1 above)
Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:418]: @TheLastProject , amended.
It now looks like this:

catima

Would you like to create a string entry for invalid balance in strings.xml?

File: app/src/main/res/values/strings.xml (hunk 3 above)
Comment (TomerPacific) [reply app/src/main/res/values/strings.xml:147]: Done.

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java (hunk 4 above)
Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java]: Ah, now I see what you meant. Changed.

**Comment by TomerPacific** [comment]
@TheLastProject , I have made some adjustments to the code and removed the logic you highlighted, but am still seeing the same tests failing.

1.     protect.card_locker.LoyaltyCardViewActivityTest > startWithoutParametersCaptureBarcodeCreateLoyaltyCard FAILED java.lang.AssertionError at LoyaltyCardViewActivityTest.java:163
2. protect.card_locker.LoyaltyCardViewActivityTest > removeBarcodeFromLoyaltyCard FAILED java.lang.AssertionError at LoyaltyCardViewActivityTest.java:163
3. protect.card_locker.LoyaltyCardViewActivityTest > startLoyaltyCardWithoutColorsSave FAILED java.lang.AssertionError at LoyaltyCardViewActivityTest.java:163
4. protect.card_locker.LoyaltyCardViewActivityTest > startLoyaltyCardWithExplicitNoBarcodeSave FAILED java.lang.AssertionError at LoyaltyCardViewActivityTest.java:163

As you can see all of them are probably failing for the same exact reason. I'm trying to investigate why.

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java (hunk 2 above)
Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java]: Removed the logic above and now only one error shows up (the one regarding the name field being empty).

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java (hunk 1 above)
Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java]: Yeah, that would be good. The old `parsingBalanceFailed` string can be remove then too as I believe this is the only place it is used.

**Comment by TheLastProject** [comment]
Hmm, I'm not quite sure why the tests are failing. But given the change still has bugs I am not very upset the tests are currently failing. The tests may become an issue later though...

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java [hunk 5]
```diff
@@ -491,6 +495,15 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
...
 
+        cardIdFieldView.setOnFocusChangeListener(new View.OnFocusChangeListener() {
+            @Override
+            public void onFocusChange(View view, boolean hasFocus) {
+                if (!hasFocus) {
+                    cardIdFieldView.setError(getString(R.string.noCardIdError));
+                }
+            }
+        });
+
```
[the remark this answers is not in the document]
Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java]: @TheLastProject  - Amended the if condition to check for the length of the string in the text view

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java (hunk 1 above)
Comment (TomerPacific) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java]: Done.

**Comment by TomerPacific** [comment]
@TheLastProject ?

**Comment by TheLastProject** [comment]
I'm not sure what you're expecting from me. The tests are failing, so this can't be merged.

**Comment by TomerPacific** [comment]
@TheLastProject,
the tests are failing because of the fact that the logic for an empty cardId has been changed in this PR.

If you take a look at **_startWithoutParametersCaptureBarcodeCreateLoyaltyCard_** test for example, you can see that:

```
 @Test
    public void startWithoutParametersCaptureBarcodeCreateLoyaltyCard() throws IOException, ParseException {
        registerMediaStoreIntentHandler();

        ActivityController activityController = Robolectric.buildActivity(LoyaltyCardEditActivity.class).create();
        activityController.start();
        activityController.visible();
        activityController.resume();

        Activity activity = (Activity) activityController.get();
        final Context context = activity.getApplicationContext();

        checkAllFields(activity,
                ViewMode.ADD_CARD,
                "",
                "",
                context.getString(R.string.anyDate),
                context.getString(R.string.never),
                "0",
                context.getString(R.string.points),
                "", //// <------------- Empty card Id
                context.getString(R.string.sameAsCardId),
                context.getString(R.string.noBarcode),
                null,
                null);

        // Complete barcode capture successfully
        captureBarcodeWithResult(activity, true);
        activityController.resume();

        checkAllFields(activity,
                ViewMode.ADD_CARD,
                "",
                "",
                context.getString(R.string.anyDate),
                context.getString(R.string.never),
                "0",
                context.getString(R.string.points),
                BARCODE_DATA,   //// <----------- Not empty cardId
                context.getString(R.string.sameAsCardId),
                BARCODE_TYPE.prettyName(),
                null,
                null);

        shadowOf(getMainLooper()).idle();

        // Save and check the loyalty card
        saveLoyaltyCardWithArguments(activity, "store", "note", context.getString(R.string.anyDate), context.getString(R.string.never), new BigDecimal("0"), context.getString(R.string.points), BARCODE_DATA, context.getString(R.string.sameAsCardId), BARCODE_TYPE.name(), true);
```
And the test fails from this assertion.

This is again, because the logic has been changed if the cardId is empty.

```
 cardIdFieldView.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean hasFocus) {
                if (!hasFocus && cardIdFieldView.getText().length() == 0) {
                    cardIdFieldView.setError(getString(R.string.noCardIdError));
                }
            }
        });
```

And,

```
 if (cardId != null) {
            cardIdFieldView.setText(cardId);
        } else {
            cardIdFieldView.setError(getString(R.string.noCardIdError));
        }
```

Since you requested these changes, let me know how you would like to proceed here.
Will the tests need to change or the logic?

**Comment by TheLastProject** [comment]
If the tests need changing after implementing the functionality because the tests turn out to not be correct but the functionality is, sure, change them.

However, from quick testing, I see some functionality is still not correct:

# Creating a card without filling in name
1. Press +
2. Scan a barcode
3. Press checkmark to save

Expected: "Card name cannot be empty" warning comes from the red icon
Actual: "Card name cannot be empty" warning shows as a toast

# Editing saved card
1. Open card
2. Press Edit

Expected: No warnings, because the card was saved so should be in a correct state
Actual: "Card ID" has a red warning which, when clicking, shows "No ID entered" despite an ID being there

It is logical that tests would fail if the behaviour is not correct yet. I'd recommend first fixing the behaviour before trying to change the tests, otherwise you may be changing tests into something that doesn't actually test the right thing.

**Pull request closed without merging**